python/taichi/main.py

```diff
@@ -1,471 +1,154 @@
-            "           ti video                  |-> Make a video using *.png files in the current folder\n"
-            "           ti video_scale            |-> Scale video resolution \n"
-            "           ti video_crop             |-> Crop video\n"
-            "           ti video_speed            |-> Speed up video\n"
-            "           ti gif                    |-> Convert mp4 file to gif\n"
-            "           ti doc                    |-> Build documentation\n"
-            "           ti release                |-> Make source code release\n"
-            "           ti debug [script.py]      |-> Debug script\n"
-            "           ti example [name]         |-> Run an example by name\n"
-        )
-        return 0
+class TaichiMain:
+    registered_commands = {
+        "example", "release", "gif", "video_speed", "video_crop",
+        "video_scale", "video", "doc", "update", "format_all", "format",
+        "build", "regression", "baseline", "benchmark", "test", "debug", "run"
+    }
 
-    t = time.time()
-    if mode.endswith('.py'):
+    @timer
+    def __init__(self, debug: bool = False, test_mode: bool = False):
+        self.banner = f"\n{'*' * 43}\n**      Taichi Programming Language      **\n{'*' * 43}"
+        print(self.banner)
+
+        if 'TI_DEBUG' in os.environ:
+            val = os.environ['TI_DEBUG']
+            if val not in ['0', '1']:
+                raise ValueError(
+                    "Environment variable TI_DEBUG can only have value 0 or 1."
+                )
+        if debug:
+            print(f"\n{'*' * 17} Debug Mode {'*' * 17}\n")
+            os.environ['TI_DEBUG'] = '1'
+
+        parser = argparse.ArgumentParser(description="Taichi CLI",
+                                         usage=self._usage())
+        parser.add_argument('command',
+                            help="command from the above list to run")
+
+        # Print help if no command provided
+        if len(sys.argv[1:2]) == 0:
+            parser.print_help()
+            exit(1)
+
+        # Parse the command
+        args = parser.parse_args(sys.argv[1:2])
+        if args.command not in self.registered_commands:
+            # TODO: do we really need this?
+            if args.command.endswith(".py"):
+                TaichiMain._exec_python_file(args.command)
+            print(f"{args.command} is not a valid command!")
+            parser.print_help()
+            exit(1)
+
+        # Flag for unit testing
+        self.test_mode = test_mode
+        if self.test_mode: return
+
+        getattr(self, args.command)(sys.argv[2:])
+
+    def _usage(self) -> str:
+        """Compose deterministic usage message based on registered_commands."""
+        # TODO: add some color to commands
+        msg = "\n"
+        space = 20
+        for command in sorted(self.registered_commands):
+            msg += f"{command}{' ' * (space - len(command))}|-> {getattr(self, command).__doc__}\n"
```

Comment:
```suggestion
            msg += f"    {command}{' ' * (space - len(command))}|-> {getattr(self, command).__doc__}\n"
```

Maybe it makes sense here to include some indentation?